velox/functions/sparksql/specialforms/AtLeastNNonNulls.cpp

```diff
@@ -0,0 +1,160 @@
+      // If 'n' > inputs_.size() - 1, result should be all false.
+      if (n > inputs_.size() - 1) {
+        return;
+      }
+    }
+    // Create a temp buffer to track count of non null values for active rows.
+    auto nonNullCounts =
+        AlignedBuffer::allocate<int32_t>(activeRows->size(), context.pool(), 0);
+    auto* rawNonNullCounts = nonNullCounts->asMutable<int32_t>();
+    for (int32_t i = 1; i < inputs_.size(); ++i) {
+      VectorPtr input;
+      inputs_[i]->eval(*activeRows, context, input);
+      if (context.errors()) {
+        context.deselectErrors(*activeRows);
+      }
+      VELOX_DYNAMIC_TYPE_DISPATCH_ALL(
+          updateResultTyped,
+          inputs_[i]->type()->kind(),
+          input.get(),
+          n,
+          context,
+          rawNonNullCounts,
+          flatResult,
+          activeRows);
+      if (activeRows->countSelected() == 0) {
+        break;
+      }
+    }
+  }
+
+ private:
+  void computePropagatesNulls() override {
+    propagatesNulls_ = false;
+  }
+
+  template <TypeKind Kind>
+  void updateResultTyped(
+      BaseVector* input,
+      int32_t n,
+      EvalCtx& context,
+      int32_t* rawNonNullCounts,
+      FlatVector<bool>* result,
+      SelectivityVector* activeRows) {
+    using Type = typename TypeTraits<Kind>::NativeType;
+    exec::LocalDecodedVector decodedVector(context);
+    decodedVector.get()->decode(*input, *activeRows);
+    bool updateBounds = false;
+    activeRows->applyToSelected([&](auto row) {
+      bool nonNull = !decodedVector->isNullAt(row);
+      if constexpr (
+          std::is_same_v<Type, double> || std::is_same_v<Type, float>) {
+        nonNull = nonNull && !std::isnan(decodedVector->valueAt<Type>(row));
+      }
+      if (nonNull) {
+        rawNonNullCounts[row]++;
+        if (rawNonNullCounts[row] >= n) {
+          updateBounds = true;
+          result->set(row, true);
+          // Exclude the 'row' from active rows after finding 'n' non-NULL /
+          // non-NaN values.
+          activeRows->setValid(row, false);
+        }
+      }
+    });
+    if (updateBounds) {
+      activeRows->updateBounds();
+    }
+  }
+};
+} // namespace
+
+TypePtr AtLeastNNonNullsCallToSpecialForm::resolveType(
+    const std::vector<TypePtr>& argTypes) {
+  VELOX_USER_CHECK_GT(
+      argTypes.size(),
+      1,
+      "AtLeastNNonNulls expect to receive at least 2 arguments.");
+  VELOX_USER_CHECK(
+      argTypes[0]->isInteger(),
+      "The first input type should be INTEGER but Actual {}.",
+      argTypes[0]->toString());
+  return BOOLEAN();
+}
+
+ExprPtr AtLeastNNonNullsCallToSpecialForm::constructSpecialForm(
+    const TypePtr& type,
+    std::vector<ExprPtr>&& compiledChildren,
+    bool /* trackCpuUsage */,
+    const core::QueryConfig& /*config*/) {
+  return std::make_shared<AtLeastNNonNullsExpr>(
+      type,
+      std::move(compiledChildren),
+      Expr::allSupportFlatNoNullsFastPath(compiledChildren));
```

Comment:
It looks like all conditions below are met for this function. Wondering if it could just be true.
```
  /// Returns true if (1) supports evaluation on all constant inputs of size >
  /// 1; (2) returns flat or constant result when inputs are all flat, all
  /// constant or a mix of flat and constant; (3) guarantees that if all inputs
  /// are not null, the result is also not null.
  virtual bool supportsFlatNoNullsFastPath() const {
    return false;
  }
```

Reply:
I think the information we need here is if all inputs support NoNullsFastPath, we can't pass true directly here.

Reply:
> if all inputs support NoNullsFastPath, we can't pass true

Could you clarify the reason? Thanks.

Reply:
You are right, I updated, thanks.